File: csharp/UAParser/OS.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace UAParser
{
  /// <summary>
  /// Represents OS information extracted from the user agent string
  /// </summary>
  public class OS
  {
    internal OS() { }
    internal OS(string family, string major, string minor, string patch, string patchMinor)
    {
      Family = family;
      Major = major;
      Minor = minor;
      Patch = patch;
      PatchMinor = patchMinor;
    }
    /// <summary>
    /// The OS family
    /// </summary>
    public string Family { get; set; }
    /// <summary>
    /// The Major version of the OS
    /// </summary>
    public string Major { get; set; }
    /// <summary>
    /// The Minor version of the OS
    /// </summary>
    public string Minor { get; set; }
    /// <summary>
    /// The patch version of the OS, if applicable
    /// </summary>
    public string Patch { get; set; }
    /// <summary>
    /// The minor patch version of the OS, if applicable
    /// </summary>
    public string PatchMinor { get; set; }

    /// <summary>
    /// Returns the string representation for the OS
    /// </summary>
    /// <returns>The OS instance as a string</returns>
    public override string ToString()
    {
      StringBuilder sb = new StringBuilder(Family);
      string version = ToVersionString();

      if (!string.IsNullOrEmpty(version))
      {
        sb.Append(' ').Append(version);
      }

      return sb.ToString();
    }
    /// <summary>
    /// The version of the OS as a single string
    /// </summary>
    /// <returns></returns>
    public string ToVersionString()
    {
      IEnumerable<string> versions = new string[] { Major, Minor, Patch, PatchMinor }.Where(x => !string.IsNullOrEmpty(x));
      return string.Join(".", versions.ToArray());
    }
  }
}

File: go/uaparser/os.go
package uaparser

import (
	"regexp"
)

type Os struct {
	Family     string
	Major      string
	Minor      string
	Patch      string
	PatchMinor string
}

type OsPattern struct {
	Regexp          *regexp.Regexp
	Regex           string
	OsReplacement   string
	OsV1Replacement string
	OsV2Replacement string
}

func (osPattern *OsPattern) Match(line string, os *Os) {
	bytes := osPattern.Regexp.FindStringSubmatch(line)
	if len(bytes) > 0 {
		groupCount := osPattern.Regexp.NumSubexp()

		if len(osPattern.OsReplacement) > 0 {
			os.Family = osPattern.OsReplacement
		} else if groupCount >= 1 {
			os.Family = bytes[1]
		}

		if len(osPattern.OsV1Replacement) > 0 {
			os.Major = osPattern.OsV1Replacement
		} else if groupCount >= 2 {
			os.Major = bytes[2]
			if len(osPattern.OsV2Replacement) > 0 {
				os.Minor = osPattern.OsV2Replacement
			} else if groupCount >= 3 {
				os.Minor = bytes[3]
				if groupCount >= 4 {
					os.Patch = bytes[4]
					if groupCount >= 5 {
						os.PatchMinor = bytes[5]
					}
				}
			}
		}
	}
}

func (os *Os) ToString() string {
	var str string
	if os.Family != "" {
		str += os.Family
	}
	version := os.ToVersionString()
	if version != "" {
		str += " " + version
	}
	return str
}

func (os *Os) ToVersionString() string {
	var version string
	if os.Major != "" {
		version += os.Major
	}
	if os.Minor != "" {
		version += "." + os.Minor
	}
	if os.Patch != "" {
		version += "." + os.Patch
	}
	if os.PatchMinor != "" {
		version += "." + os.PatchMinor
	}
	return version
}
